clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -246,17 +228,25 @@ public Selector(long connectionMaxIdleMS, int failedAuthenticationDelayMs, Metri
      * @throws IOException if DNS resolution fails on the hostname or if the broker is down
      */
     @Override
-    public void connect(String id, InetSocketAddress address, int sendBufferSize, int receiveBufferSize) throws IOException {
-        ensureNotRegistered(id);
+    public void connect(Node node, int sendBufferSize, int receiveBufferSize) throws IOException {
```

Comment:
Because the client calls `connect` which then hands off to NIO the asynchronous connection 
and it's in handling the keys that we can see if one IP was "bad" and all this without the client being involved at all.

It may be not the smallest change possible because of the footprint in the tests, but to us this was the change that seemed to make sense.

We are biased by the fact that the existing `Selectable.connect` signature wasn't great also for KIP-81

Reply:
Network layer code is already rather complex and I would avoid this type of client logic into this layer unless there was no other way. For disconnections, we already propagate channel state to clients and it would be quite easy to add another state to indicate that it may want to retry. A few advantages of doing it this way:
1. Keep network layer changes to minimum
2. We want to make this change only for clients, so the keeping the change in the client layer makes sense (making it easier to read the code when debugging broker issues)
3. Reduces the footprint of this change
4. Keeps the handling of different variations of this config in the same layer (since the other values of this config are handled in the client layer)

I haven't followed KIP-81 PR discussions, so if there are other reasons to make this change in `Selector`, then perhaps it may make sense. But just looking at this one, it feels like it should be the responsibility of `NetworkClient` to iterate through the addresses.

@hachikuji Since you reviewed the KIP-81 PR, what do you think?

Reply:
Many thanks @rajinisivaram for taking a look under heavy load!
We see the point of not polluting the Network layer and having retries done by the client only.